guides/api-style.rst

```diff
@@ -0,0 +1,118 @@
+* Interacting arguments (e.g. arguments that must be in a particular order, or
+  where at most one is valid, or where one argument restricts the valid range
+  of the other) are fine, but when this happens the behaviour of defaults
+  should automatically be adjusted. e.g. if the normal default of an argument
+  would become invalid, the function should still do the right thing if that
+  default is used.
+* Where the actual default used depends on other arguments, the default parameter
+  should be None.
+* It's worth thinking about the order of arguments: the first one or two
+  arguments are likely to be passed positionally, so try to put values there
+  where this is useful and not too confusing.
+* Arguments should not be "a value or a strategy for generating that value"
+  unless there is some highly compelling reason for that. Hypothesis has good
+  composition mechanisms and people should use them instead of adding them on
+  an ad hoc basis to individual API functions.
+
+~~~~~~~~~~~~~~
+Function Names
+~~~~~~~~~~~~~~
+
+We don't have any real consistency here. The rough approach we follow is:
+
+* Names are `snake_case` as is standard in Python.
+* Strategies for a particular type are typically named as a plural name for
+  that type. Where that type has some truncated form (e.g. int, str) we use a
+  longer form name.
+* Other strategies have no particular common naming convention.
+
+~~~~~~~~~~~~~~
+Argument Names
+~~~~~~~~~~~~~~
+
+We should try to use the same argument names and orders across different
+strategies wherever possible. In particular:
+
+* For collection types, the element strategy (or strategies) should always be
+  the first argument. It does not have a specific common name.
+* For ordered types, the first two arguments should be a lower and an upper
+  bound. They should be called ``min_value`` and ``max_value``.
+* Collection types should have a ``min_size`` and a ``max_size`` parameter that
+  controls the range of their size. ``min_size`` should default to zero and
+  ``max_size`` to ``None`` (even if internally it is bounded).
+
+
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+A catalogue of current violations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following are places where we currently deviate from this style. Some of
+these should be considered targets for deprecation and/or improvement.
+
+* most of the collections in ``hypothesis.strategies`` have an ``average_size``
+  distribution hint.
+* many of the collections in ``hypothesis.strategies`` allow a default of
+  ``None`` for their elements strategy (meaning only generate empty
+  collections).
+* the date and time handling strategies in ``hypothesis.strategies`` has
+  bound names that match the strategy name rather than being ``min_value`` and
+  ``max_value``
+* ``hypothesis.extra.numpy`` has a lot of arguments which can be either
+  strategies or values.
+* ``hypothesis.extra.numpy`` assumes arrays are fixed size and doesn't have
```

Comment:
> IMO min_size and max_size only really work for one-dimensional collections - for N-dimensional arrays

Yeah. I don't necessarily think deviating from the style was a *mistake* here, but it's worth noting as a deviation.

>  I think allowing exact size arguments only and pushing flatmap is too much of a haskellism 

I'd generally push `composite()` or `data()` over `flatmap` really. Python makes `flatmap` too needlessly cumbersome because of the lack of multiline lambdas.

A thing that hasn't really propagated through to the AP/documentation in any meaningful way but may at some point if I figure out how to make it play well in Python is to start pushing the use of data a bit more. In other language's versions of Hypothesis the primary interface is going to look more like `@given(data())` and will lack most of the fancy argument wrangling.